mozilla_version/fenix.py

```diff
@@ -0,0 +1,77 @@
+"""Defines characteristics of a Fenix version at Mozilla."""
+
+import attr
+import re
+
+from mozilla_version.version import BaseVersion
+from mozilla_version.parser import strictly_positive_int_or_none
+
+
+@attr.s(frozen=True, cmp=False, hash=True)
+class FenixVersion(BaseVersion):
+    """Class that validates and handles Fenix version numbers."""
+
+    _VALID_ENOUGH_VERSION_PATTERN = re.compile(r"""
+        ^(?P<major_number>\d+)
+        \.(?P<minor_number>\d+)
+        (\.(?P<patch_number>\d+))?
+        (
+            -beta\.(?P<beta_number>\d+)
+            |-rc\.(?P<release_candidate_number>\d+)
+        )?$""", re.VERBOSE)
+
+    _OPTIONAL_NUMBERS = BaseVersion._OPTIONAL_NUMBERS + (
+        'beta_number', 'release_candidate_number'
+    )
+
+    _ALL_NUMBERS = BaseVersion._ALL_NUMBERS + _OPTIONAL_NUMBERS
+
+    beta_number = attr.ib(type=int, converter=strictly_positive_int_or_none, default=None)
+    release_candidate_number = attr.ib(
+        type=int, converter=strictly_positive_int_or_none, default=None
+    )
+
+    @classmethod
+    def parse(cls, version_string):
+        """Construct an object representing a valid Fenix version number."""
+        return super(FenixVersion, cls).parse(version_string)
+
+    @property
+    def is_beta(self):
+        """Return `True` if `FenixVersion` was built with a string matching a beta version."""
+        return self.beta_number is not None
+
+    @property
+    def is_release_candidate(self):
+        """Return `True` if `FenixVersion` was built with a string matching an RC version."""
+        return self.release_candidate_number is not None
+
+    def __str__(self):
+        """Implement string representation.
+
+        Computes a new string based on the given attributes.
+        """
+        string = super(FenixVersion, self).__str__()
+
+        if self.is_beta:
+            string = '{}-beta.{}'.format(string, self.beta_number)
+        elif self.is_release_candidate:
+            string = '{}-rc.{}'.format(string, self.release_candidate_number)
+
+        return string
+
+    def _compare(self, other):
+        if isinstance(other, str):
+            other = FenixVersion.parse(other)
+        elif not isinstance(other, FenixVersion):
+            raise ValueError('Cannot compare "{}", type not supported!'.format(other))
+
+        difference = super(FenixVersion, self)._compare(other)
+        if difference != 0:
+            return difference
+
+        # XXX: Fix before landing:
```

Comment:
👍 

FWIW, Gecko implements this logic at:

https://github.com/mozilla-releng/mozilla-version/blob/2180fe058bc3ada6fa8cbdd24cc847f27b418798/mozilla_version/gecko.py#L340-L348